### DISCUSSION TITLE
Custom font with company-mode

### DISCUSSION THREAD

**Discussion by umgefahren** [company-mode/company-mode#1364] [Q&A]
Without setting my custom font (JetBrains Mono), company mode looks correct. However, if I set my font with this:
```elisp
(add-to-list 'default-frame-alist '(font . "JetBrains Mono-12"))
```
completions look like this:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/55623006/213774627-1bb91fc9-5fd0-43fa-8862-838b6bb24a94.png">
The icons are way smaller now. How do I correct that?

I'm very new to Emacs so forgive me if my question is stupid.

This is my complete config:
https://gist.github.com/umgefahren/303f9df36799cb7f68f62f599572b30f

**Comment by dgutov** [comment]
Hmm, what is your value of `company-icon-margin`? What happens if you increase it to 2 or 3?

**Reply by umgefahren** [reply]
This is how it looks like with bigger icon margins:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/55623006/213917768-563f35a7-9470-4417-a45a-c85ee0644a55.png">

**Reply by dgutov** [reply]
All right. It seems like this font has a lower height (of the characters themselves) than the previous font you used.

The value to check is what is returned by the form `(default-font-height)`.

If so, you can try specifying the icon size explicitly, using the variable `company-icon-size`.

**Reply by dgutov** [reply]
Or! You can try this patch:

```diff
diff --git a/company.el b/company.el
index af056eb..5691d58 100644
--- a/company.el
+++ b/company.el
@@ -1618,7 +1618,7 @@ end of the match."
                          ((and (consp company-icon-size)
                                (eq 'auto-scale (car company-icon-size)))
                           (let ((base-size (cdr company-icon-size))
-                                (dfh (default-font-height)))
+                                (dfh (default-line-height)))
                             (min
                              (if (> dfh (* 2 base-size))
                                  (* 2 base-size)
```

**Reply by umgefahren** [reply]
> All right. It seems like this font has a lower height (of the characters themselves) than the previous font you used.
> 
> The value to check is what is returned by the form `(default-font-height)`.
> 
> If so, you can try specifying the icon size explicitly, using the variable `company-icon-size`.

I tried that.

Adding
```elisp
(setq company-icon-size (default-font-height))
```
to my config section in use-package only increases the width of the margin.

**Reply by dgutov** [reply]
I meant try a different size. A bigger one.

Anyway, better try the patch first.

**Reply by dgutov** [reply]
Actually, never mind that patch. I just committed a more sensible tweak.

It might not help you, though, if your font is even smaller.

Just try setting `company-icon-size` to different (integer) values and see which looks right to you.

The source of the problem is that you probably have some display scaling enabled. And while there we do make an effort to perform some automatic scaling, these svg-s look best at certain sizes that are multiple of the original (which is 16). So I'm only doing automatic 2x scaling if it fits, which in your case it apparently does not.

**Reply by dgutov** [reply]
We can do auto-scaling to font height as a separate value, though:

```diff
diff --git a/company.el b/company.el
index 11f8b06..96b3899 100644
--- a/company.el
+++ b/company.el
@@ -1623,7 +1623,10 @@ end of the match."
                              (if (>= dfh (* 2 base-size))
                                  (* 2 base-size)
                                base-size)
-                             (* company-icon-margin dfw))))))
+                             (* company-icon-margin dfw))))
+                         ((eq company-icon-size 'auto-scale)
+                          (min (default-font-height)
+                               (* company-icon-margin dfw)))))
              (spec (list 'image
                          :file (expand-file-name icon-file root-dir)
                          :type 'svg
```

You can probably see that the icons look a little imperfect, though.
